Monorepo: add package.json linter

simison opened the pull request:
Lint monorepo `package.json` files.

Uses neat [`npm-package-json-lint`](https://github.com/tclindner/npm-package-json-lint) for the job.

Linter config based on [`@wordpress/npm-package-json-lint-config`](https://www.npmjs.com/package/@wordpress/npm-package-json-lint-config) and customized a bit to our needs, with a few overrides.

#### Changes proposed in this Pull Request

* Adds `package.json` linter, not only to keep monorepo packages consistent but especially to keep `devDependencies` out from `packages/**/package.json` since they can cause actual issues.

#### Testing instructions

* Run `npm run lint:package-json`

Linter runs as part of `npm run lint` as well.

simison committed:
Monorepo: add package.json linter

simison committed:
docs

simison commented on docs/monorepo.md:
```diff
@@ -51,7 +51,7 @@ The only exception are `devDependencies` which _must be declared in the wp-calyp
 	"sideEffects": false,
 	"keywords": [ "wordpress" ],
-	"author": "Your Name <you@example.com> (https://yoursite.wordpress.com/)",
+	"author": "Automattic Inc.",
```
I think this is better so as not to promote too tight "code ownership".

I can ping individuals who's names I've switched to the company name in this PR once it's in a mergeable state.

simison committed:
npm-package-json-lint@4.0.1

simison committed:
Handle i18n-calypso-cli

simison committed:
Sort out config

simison commented on npmpackagejsonlint.config.js:
```diff
@@ -0,0 +1,49 @@
+module.exports = {
+	// See https://www.npmjs.com/package/@wordpress/npm-package-json-lint-config
+	extends: '@wordpress/npm-package-json-lint-config',
```
Alternatively, just list all those rules here and have one package less to Renovate/update. ;-)

sirreal replied:
🤷‍♂ How much value do we get value from extending? I'm inclined to just list our own rules here, it's pretty trivial to pull in or remove a rule when it makes sense, while keeping the package up to date adds overhead and it can be surprising when rules shift around.

simison replied:
>How much value do we get value from extending?

Whenever anything changes/gets deprecated in the core, we'd inherit those changes. We have that problem with our Stylelint config - a bunch of it was outdated last time I checked.

simison committed:
4.0.1 → 4.0.2

simison commented:
@gziolo does it make sense to use `@wordpress/npm-package-json-lint-config` like this?

PS. might be new [`prefer-no-devdependencies`](https://npmpackagejsonlint.org/docs/en/rules/disallowed-nodes/prefer-no-devdependencies) rule in `npm-package-json-lint` v4 would be useful to Gutenberg repo as well?

gziolo commented:
> @gziolo does it make sense to use `@wordpress/npm-package-json-lint-config` like this?

Sure, this is how we use it internally in Gutenberg as well:
https://github.com/WordPress/gutenberg/blob/4df25af5946c518d46ebe14a24aea1d44cd63593/package.json#L158

> PS. might be new [`prefer-no-devdependencies`](https://npmpackagejsonlint.org/docs/en/rules/disallowed-nodes/prefer-no-devdependencies) rule in `npm-package-json-lint` v4 would be useful to Gutenberg repo as well?
Cool, see https://github.com/WordPress/gutenberg/pull/18054 :)

gziolo commented on npmpackagejsonlint.config.js:
```diff
@@ -0,0 +1,49 @@
+		'require-keywords': 'off',
+		'require-license': 'error',
+		'require-name': 'error',
+		'require-version': 'error',
```
I see the in the WP config already.

gziolo commented on npmpackagejsonlint.config.js:
```diff
@@ -0,0 +1,49 @@
+	rules: {
+		'prefer-no-devDependencies': 'error',
+		'prefer-property-order': 'off',
+		'require-author': 'error',
```
I see it in the WP config already.

gziolo commented on npmpackagejsonlint.config.js:
```diff
@@ -0,0 +1,49 @@
+		'require-name': 'error',
+		'require-version': 'error',
+		'valid-values-author': [ 'error', [ 'Automattic Inc.' ] ],
+		'valid-values-license': [ 'error', [ 'GPL-2.0-or-later' ] ],
```
This one is also present in the shared config, although you might want to be more opinionated.

simison replied:
Removed this and other redundant ones. https://github.com/Automattic/wp-calypso/pull/36534/commits/9db72d1c7ae01192e004c799a390cf96642cb7bd

gziolo commented on package.json:
```diff
@@ -255,6 +256,7 @@
 		"lint:css": "stylelint \"**/*.scss\" --syntax scss",
 		"lint:js": "npm run -s install-if-deps-outdated && eslint --ext .js --ext .jsx --ext .ts --ext .tsx --cache .",
 		"lint:mixedindent": "mixedindentlint --ignore-comments \"client/**/*.scss\" \"assets/**/*.scss\" \"**/*.js\" \"**/*.jsx\" \"**/*.tsx\" \"**/*.tsx\" \"!build/**\" \"!node_modules/**\" \"!public/**\" \"!client/config/index.js\"",
+		"lint:package-json": "npmPkgJsonLint ./package.json ./packages/**/package.json ./apps/**/package.json",
```
You probably should wrap patterns with quotes to prevent the default behaviors of CLIs. When using ZSH, it will convert those patterns into a list of matched files. I guess it's still correct but each env can behave differently.

simison replied:
yup! https://github.com/Automattic/wp-calypso/pull/36534/commits/3b409a5de1eb97b64dc07279b4caed04dd6f17ce

sirreal committed:
Alphabetize dependencies

sirreal commented:
Rebased

simison committed:
Add quotes to globs

sirreal commented on package.json:
```diff
@@ -255,6 +256,7 @@
 		"lint:css": "stylelint \"**/*.scss\" --syntax scss",
 		"lint:js": "npm run -s install-if-deps-outdated && eslint --ext .js --ext .jsx --ext .ts --ext .tsx --cache .",
 		"lint:mixedindent": "mixedindentlint --ignore-comments \"client/**/*.scss\" \"assets/**/*.scss\" \"**/*.js\" \"**/*.jsx\" \"**/*.tsx\" \"**/*.tsx\" \"!build/**\" \"!node_modules/**\" \"!public/**\" \"!client/config/index.js\"",
+		"lint:package-json": "npmPkgJsonLint './package.json' './packages/**/package.json' './apps/**/package.json'",
```
These globs could likely be "smaller":

```suggestion
		"lint:package-json": "npmPkgJsonLint './package.json' './packages/*/package.json' './apps/*/package.json'",
```

We'd expect to match `packages/foo/package.json`, but not `packages/an/arbitrary/directory/depth/package.json`.

simison committed:
4.0.2 → 4.0.3

simison committed:
Fix lint error

simison committed:
Remove duplicates already in config we extend

https://github.com/WordPress/gutenberg/blob/3febc48d8240a0f8b04b11201a30178b14534c5d/packages/npm-package-json-lint-config/index.js

simison committed:
Update package.json

Co-Authored-By: Jon Surrell <jon.surrell@automattic.com>

blowery commented on npmpackagejsonlint.config.js:
```diff
@@ -0,0 +1,44 @@
+	overrides: [
+		{
+			patterns: [
+				'./apps/full-site-editing/package.json',
```
any reason these are explicit instead of `./apps/*/package.json`?

simison replied:
No other than each app is a bit different and not all of them have package-lock currently. Happy to have it more generic. 👍

blowery replied:
do they need a lock to be linted?

blowery replied:
fine for now

blowery commented on npmpackagejsonlint.config.js:
```diff
@@ -0,0 +1,44 @@
+		{
+			patterns: [
+				'./package.json',
+				'./packages/eslint-config-wpcalypso/package.json',
```
ditto for these. could we use a * to match all of the packages?

simison replied:
I would expect `valid-values-name-scope` rule to be enforced by default and this is the list of packages that are old and didn't follow `@automattic/*` naming convention.

No new packages should be added without `@automattic/*`, right?

blowery replied:
ahhh, of course. 👍

blowery reviewed (approved):
I think this is good to land. Anything else you want to add?

sirreal reviewed (approved):
:shipit:
